In [1]:
# importing libraries
import pandas as pd
import numpy as np
import os

In [2]:
path = r'/Users/priyankamittal/Desktop/Data Analytics/CareerFoundry/Immersion/Python/07-2023 Instacart Basket Analysis'

In [3]:
path

'/Users/priyankamittal/Desktop/Data Analytics/CareerFoundry/Immersion/Python/07-2023 Instacart Basket Analysis'

In [27]:
df_prods = pd.read_csv(os.path.join(path,'Data','Orginal Data','products.csv'),index_col = False)

In [28]:
df_ords = pd.read_csv(os.path.join(path,'Data','Orginal Data','orders.csv'),index_col = False)

In [29]:
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [30]:
df_prods.tail()

product_id                                product_name  aisle_id  \
49688       49684   Vodka, Triple Distilled, Twist of Vanilla       124   
49689       49685          En Croute Roast Hazelnut Cranberry        42   
49690       49686                            Artisan Baguette       112   
49691       49687  Smartblend Healthy Metabolism Dry Cat Food        41   
49692       49688                      Fresh Foaming Cleanser        73   

       department_id  prices  
49688              5     5.3  
49689              1     3.1  
49690              3     7.8  
49691              8     4.7  
49692             11    13.5

In [31]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [14]:
#creating a data frame for column have mixed data type
df_test =pd.DataFrame()

In [16]:
df_test['mix']= ['a','b','1','True']

In [17]:
df_test

mix
0     a
1     b
2     1
3  True

In [19]:
# check for mix data type in existing table
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

In [20]:
col

'mix'

In [21]:
# change the datatype for entry '1' to str
df_test['mix'] = df_test['mix'].astype('str')

In [22]:
df_test.dtypes

mix    object
dtype: object

In [23]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

In [24]:
col

'mix'

In [25]:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   mix     4 non-null      object
dtypes: object(1)
memory usage: 160.0+ bytes


## Missing Values

In [32]:
# check any Nan value for entire datafram(isnull() and sum is used for total entries in aprticular column which is having nan value)
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [33]:
# Creating a subset for only NaN values
df_nan = df_prods[df_prods['product_name'].isnull()==True]

In [34]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [37]:
#fill in with mean value df['column with missings'].fillna(mean value, inplace=True)
df_prods.mean()

/var/folders/m8/p9jl94914t5_8x2_kf93jm240000gn/T/ipykernel_10222/730692063.py:1: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  df_prods.mean()


product_id       24844.345139
aisle_id            67.770249
department_id       11.728433
prices               9.994136
dtype: float64

In [38]:
#fill in with median value df['column with missings'].fillna(median value, inplace=True)
df_prods.median()

/var/folders/m8/p9jl94914t5_8x2_kf93jm240000gn/T/ipykernel_10222/3102030757.py:2: FutureWarning: The default value of numeric_only in DataFrame.median is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  df_prods.median()


product_id       24845.0
aisle_id            69.0
department_id       13.0
prices               7.1
dtype: float64

In [40]:
df_prods.shape

(49693, 5)

In [41]:
# Since the column type is a string so can't use mean or median to add value,so will delete those data
df_prods_clean = df_prods[df_prods['product_name'].isnull()==False]

In [42]:
df_prods_clean.shape

(49677, 5)

In [ ]:
#Another way you can drop all missing values is via the following command:

# df_prods.dropna(inplace = True)
# If you wanted to use this command to drop only the NaNs from a particular column, the code would look like this:

# df_prods.dropna(subset = [‘product_name’], inplace = True)

## Duplicate values

In [44]:
# Identifying and storing duplicate values on df_dups
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [45]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [46]:
df_prods_clean.shape

(49677, 5)

In [49]:
df_prods_clean_no_dup = df_prods_clean.drop_duplicates()

In [50]:
df_prods_clean_no_dup.shape

(49672, 5)

In [53]:
df_prods_clean_no_dup.to_csv(os.path.join(path,'Data','Prepared Data','products_checked.csv'))

## Task 2: Validating outlier in order dataframe
Tip: Keep an eye on min and max values!

In [54]:

df_ords.describe()

# There is no outlier 

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

## Validating product dataframe

In [55]:
df_prods_clean_no_dup.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

#### The max value of "prices" seems irrelevant, it's even too high compared to the 3rd quartile (75%)

In [56]:
df_prods_clean_no_dup.loc[df_prods_clean_no_dup['prices']==99999]

product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0

### Step 3 (Mixed-type data)

In [64]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col,' mixed')
    else: print (col,' consistent')

SyntaxError: invalid syntax (979248044.py, line 5)

In [58]:
col

'days_since_prior_order'

In [60]:
# Check for mixed-type data in your df_ords dataframe.

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col, ' mixed')
  else: print (col, ' consistent')

order_id  consistent
user_id  consistent
eval_set  consistent
order_number  consistent
order_dow  consistent
order_hour_of_day  consistent
days_since_prior_order  consistent


In [65]:
df_ords.dtypes

order_id                    int64
user_id                     int64
eval_set                   object
order_number                int64
order_dow                   int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

### Step 4

# If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

#### No mixed-data found

### Step 5 (Missing values)

In [66]:
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [72]:
# Create subset for the missing values in df_ords

df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [73]:
df_ords_nan

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
11        2168274        2    prior             1          2   
26        1374495        3    prior             1          1   
39        3343014        4    prior             1          6   
45        2717275        5    prior             1          3   
...           ...      ...      ...           ...        ...   
3420930    969311   206205    prior             1          4   
3420934   3189322   206206    prior             1          3   
3421002   2166133   206207    prior             1          6   
3421019   2227043   206208    prior             1          1   
3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [74]:
df_ords[df_ords['user_id'] == 1]

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0    2539329        1    prior             1          2                  8   
1    2398795        1    prior             2          3                  7   
2     473747        1    prior             3          3                 12   
3    2254736        1    prior             4          4                  7   
4     431534        1    prior             5          4                 15   
5    3367565        1    prior             6          2                  7   
6     550135        1    prior             7          1                  9   
7    3108588        1    prior             8          1                 14   
8    2295261        1    prior             9          1                 16   
9    2550362        1    prior            10          4                  8   
10   1187899        1    train            11          4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

In [75]:
df_ords[df_ords['user_id'] == 2]

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
11   2168274        2    prior             1          2                 11   
12   1501582        2    prior             2          5                 10   
13   1901567        2    prior             3          1                 10   
14    738281        2    prior             4          2                 10   
15   1673511        2    prior             5          3                 11   
16   1199898        2    prior             6          2                  9   
17   3194192        2    prior             7          2                 12   
18    788338        2    prior             8          1                 15   
19   1718559        2    prior             9          2                  9   
20   1447487        2    prior            10          1                 11   
21   1402090        2    prior            11          1                 10   
22   3186735        2    prior            12          1                  9   
23   3268552        2    prior            13          4                 11   
24    839880        2    prior            14          3                 10   
25   1492625        2    train            15          1                 11   

    days_since_prior_order  
11                     NaN  
12                    10.0  
13                     3.0  
14                     8.0  
15                     8.0  
16                    13.0  
17                    14.0  
18                    27.0  
19                     8.0  
20                     6.0  
21                    30.0  
22                    28.0  
23                    30.0  
24                    13.0  
25                    30.0

In [76]:
df_ords[df_ords['user_id'] == 3]

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
26   1374495        3    prior             1          1                 14   
27    444309        3    prior             2          3                 19   
28   3002854        3    prior             3          3                 16   
29   2037211        3    prior             4          2                 18   
30   2710558        3    prior             5          0                 17   
31   1972919        3    prior             6          0                 16   
32   1839752        3    prior             7          0                 15   
33   3225766        3    prior             8          0                 17   
34   3160850        3    prior             9          0                 16   
35    676467        3    prior            10          3                 16   
36    521107        3    prior            11          0                 18   
37   1402502        3    prior            12          1                 15   
38   2774568        3     test            13          5                 15   

    days_since_prior_order  
26                     NaN  
27                     9.0  
28                    21.0  
29                    20.0  
30                    12.0  
31                     7.0  
32                     7.0  
33                     7.0  
34                     7.0  
35                    17.0  
36                    11.0  
37                    15.0  
38                    11.0

In [77]:
df_ords[df_ords['user_id'] == 4]

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
39   3343014        4    prior             1          6                 11   
40   2030307        4    prior             2          4                 11   
41    691089        4    prior             3          4                 15   
42     94891        4    prior             4          5                 13   
43   2557754        4    prior             5          5                 13   
44    329954        4     test             6          3                 12   

    days_since_prior_order  
39                     NaN  
40                    19.0  
41                    21.0  
42                    15.0  
43                     0.0  
44                    30.0

#### Through the two examples above I can tell that any user who has an order number of "1" will have "NaN" as their days_since_prior_order. This make sense since it was their first order, so they wouldn't have any prior orders and that's why they won't have any data showing how long it has been since the previous order. Maximum user id is 206209 and total missing values for days_since_prior_order is also 206209, meaning each user will have one NaN value located in their first order.

#### I decided not to make any changes. Even though it looks missing, actually the data itself has value in it. We can use it to show that it's actually their first order they've ever made or they're a new customer who just made their first order.

In [ ]:
### Step 7 (Duplicates)

In [80]:
df_ords_dup = df_ords[df_ords.duplicated()]

In [81]:
df_ords_dup

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

In [83]:
df_ords_dup.size

0

#### No duplicate found`

### Step 8 (Addressing duplicates)


#### Since there is no duplicate found so no addressing of suplicate is required.
####### Some duplicates may occur multiple times in one column. Other times, you may have full duplicates. In these cases, you’ll have multiple rows with the exact same values.

In both of these cases, you’ll need to make a record of these occurrences, export examples, and send them to your client for further explanation. Oftentimes, the duplicates you encounter won’t make sense to you, so you’ll need clarification from your client before proceeding with any data manipulation. This can also be handy for your clients—they may not have realized the data contains duplicates. Having these flagged can allow them to identify problems in their data collection and storage processes.



In [84]:
df_ords_dup.to_csv(os.path.join(path,'Data','Prepared Data','orders_checked.csv'))